### ISSUE TITLE
.put not saving/changing anything

### ISSUE THREAD

**Issue by Napzu** [amark/gun#426]
**Demoed and explained in detail here:** https://www.kotisivu.ninja/gunbug2

Gun version used: 0.8

localstorage is cleared every page load before gun initialization

GunDB is connected to server

testObj = website.gun.get('gunBugRootkey').get('testObj');

**BUGS**
After pageload/localStorage.clear(), if initial value of testObj is:
{ 1: 'SIMPLE VAL' },                   you cannot change it to { 1: { 'something' : 'qwerty' } }
{ 1: { 'something' : 'qwerty' } }, you cannot change it to { 1: 'SIMPLE VAL' }
{ 1: 'SIMPLE VAL' },                   you cannot change it to { 1: { PUT ANYTHING HERE } }
{ 1: { 'something' : 'qwerty' } }, you cannot change it to { 1:  'PUT ANYTHING HERE'  }

If you null testObj first, then you can change the value to whichever

**Also really strange bug**, after pageload, if initial value is { 1: 'SIMPLE VAL' }
Try to change it **TWO TIMES** to { 1: { 'something' : 'qwerty' } }, ( log reports that it still is "SIMPLE VAL" )
Change back to null
Now you cannot change it back to { 1: { 'something' : 'qwerty' } } anymore????

**Comment by amark** [comment]
Thanks for reporting, this is going into the queue of fixes that need to be made.

**Comment by d3x0r** [comment]
Client gets all events correctly... server doesn't get any events. (except on second run with non blank data.json)  This sample does end up using the same file for client and server.
```

var Gun = require( "gun" );
var port = process.env.OPENSHIFT_NODEJS_PORT || process.env.VCAP_APP_PORT || process.env.PORT || process.argv[2] || 8080;

var server = require('http').createServer(function(req, res){
    var insert = "";
    console.log( "Localhost connect?" );
    if( req.url.endsWith( "gun.js" ) )
        insert = "../";

    require('fs').createReadStream(require('path').join(__dirname, insert, req.url)).on('error',function(){ // static files!
        res.writeHead(200, {'Content-Type': 'text/html'});
        res.end(require('fs').readFileSync(require('path').join(__dirname, 'index.html'))); // or default to index
    }).pipe(res); // stream
});
server.listen(port);

//var gunServer = Gun({ 
//    ws: { servers: [server] }
//});
var gunServer = Gun({ 
    web: server
});
gunServer.get("root").map().on( (val,field)=>{
console.log( "svr:root:", field,"=",val );
});

gunServer.get("root").get('notRoot').map().on( (val,field)=>{
console.log( "svr:root.notRoot:", field,"=",val );
});

var gunClient = new Gun( "localhost:"+port );
var db = gunClient.get( "root" ).get ("notRoot" );

gunClient.get("root").map().on( (val,field)=>{
console.log( "cli:root:", field,"=",val );
});

db.map().on( (val,field)=>{
console.log( "cli:root.notRoot:", field,"=",val );
});

db.put( 3 );
db.put( {field:4} );

db.put( 4 );
db.put( {field:5} );

db.put( 5 );
```

**Comment by amark** [comment]
@Napzu I should have made a copy of your demo page, because it is currently being reported as offline/down/not-available/cannot-be-found/not-reachable.

@Napzu I am guessing that `v0.8.8` has now fixed this problem (it has already fixed some of the other issues reported), however I wasn't able to use your example page to test if it is fixed. Could you please check after upgrading?

In my test, I was able to:

 1. Change { 1: 'SIMPLE VAL' } to { 1: { 'something' : 'qwerty' } }.
 2. Change { 1: { 'something' : 'qwerty' } } to { 1: 'SIMPLE VAL' }.
 3. And so on...

I'll give this a couple days to hear back from you to verify that what I got working is also working for you. If I don't hear back, I'll go ahead and close the issue as fixed.

Likewise, @d3x0r when I run your code in node on the latest, the last console.log I get is `notRoot = 5` which I believe is the intended/expected result since `5` is the last thing you save. Even if I re-run it the second time, it produces the same correct output.

Let me know! Thanks for reporting this!

**Comment by Napzu** [comment]
Damn, forget to renew letsencrypt cert.

Now https://www.kotisivu.ninja/gunbug2 is working again. On my test page, same problems still exists.

**Comment by d3x0r** [comment]
Yes in my test cases they were working as expected.. but I was only doing it locally.  I think it's something to do with the ws driver that's causing the issue.
(also with inability to clear a node with `null` only shows up with ws)

**Comment by megaed** [comment]
I have the same issue. 
Put not saving changes at all...

I'm currently using the latest version of GunJS.

**Comment by amark** [comment]
@megaed thanks for also reporting this.

I'm having a hard time understanding what if (A) I'm getting the correct output because I'm running the example wrong (B) If I'm getting the wrong output but I don't know that it is wrong.

Could any one of you, @megaed , @Napzu , @d3x0r do me a favor and convert what you are saying into a mocha test? Here is an example:

```
it('check value is what we saved it as', function(done){
	var gun = Gun();
	gun.get('key').get('hello').put('world');

	gun.get('key').get('hello').val(function(data){
		expect(data).to.be('world');
		done();
	});
});
```

This will help me out, thanks!

**Comment by jonmalave** [comment]
I reverted back to v0.8.7 due to put issue in v0.8.8. Here is the code snippet of what I am using to reproduce the issue: 

```
user: gun.get('members').get('account'),
userId: '_1',
chat: gun.get('chat'),
send({ content, media }) {
  this.user.get(this.userId).get('msgId').val((data) => {
     const id = ++data;
     this.user.get(this.userId).get('msgId').put(id);
     this.chat.get('message').put({
         [`user${this.userId}msg${id}`]: {
         author: store.user.name,
         profile: store.user.profile,
         content,
         media
     }
    });
 });
}
```

the above code is what i am currently using to get the current user of my Chat App, once we have the id of the current user we are then appending their new chat message to a `chat.message` node with the `send` method. In the send method we grab the current users messageId (which is just my way of keeping count of number of messages user sent, and also used as a chat message id that is appended to the chat message's `KEY` it becomes the unique message id combined with their userId. 

I am able to process the above with no issue on v0.8.7 but not in v0.8.8. Basically in v0.8.8 the first message is saved, but each subsequent message sent is never saved so the `chat.message` node never is able to receive additional appended messages (data) except for the first one.

**Comment by Napzu** [comment]
I am new to mocha test. Can mocha test made headless on browser side? Because bugs are on the browser side with connection to gun server.

**Comment by amark** [comment]
@jonmalave thanks for chiming in. This is probably my "next up" thing to fix, but unfortunately your code wasn't helpful for replicating :(.

@Napzu yes, this is the template that will let you do browser-side mocha testing (if you need to test complex interactions between server/browser then you may need PANIC, but hopefully not) here:

https://github.com/amark/gun/blob/master/test/mocha.html

You can probably just load those files into a jsbin from jsdelivr, and then all you have to do is add another script tag with your test in it, here is some quick pseudocode on how to write a mocha test:

```
describe('foo', => {
  it('bar', done => {
    expect(a).to.be('lol');
    done();
  });
});
```

^^ or something like that.

**Comment by amark** [comment]
@napzu @d3x0r I think we figured it out!!!

Somebody else reported a different bug on StackOverflow (instead of GitHub) here: https://stackoverflow.com/questions/46937247/nested-object-is-not-updated

I was able to correctly replicate it (sorry I wasn't able to figure out how to replicate yours), and got it fixed this morning. Once I realized what was going on, I'm 99.99999% this is the same bug as this one, because this one uses a nested put also!

Remember, I had previously written and tested a fix that I thought was replicating your guys issue, but the difference was being 1 layer deeper than what I was checking. It made all the difference.

I'm highly confident this is finally fixed now!!! (and am gonna close this issue, definitely reopen if necessary)

I am so sorry this took forever @napzu and delayed/postponed so many of your awesome/cool projects. I hope this lets you get started back up on them again! I'm looking forward to seeing what you build (and lol, any new bugs you find).

@d3x0r thanks a ton for your debugging as well, you've been so helpful with a lot of these, and I've really appreciated it, it has taken a lot of stress off me.

**Comment by amark** [comment]
@Napzu is there any way I can get you to console.log("NOT WORKING") on every wrong action, and console.log("WORKS!") on any correct action? Or better, a mocha test? I provided the framework for how to write one in one of my other above comments.

**Comment by Napzu** [comment]
Mocha or something else you to better test this is totally in my todo list. I have so many client projects at the moment... But I try to make it this weekend.

**Comment by amark** [comment]
👍 thanks @Napzu !

**Comment by Napzu** [comment]
Sorry I haven't got time to try out mocha test..

But I updated my https://www.kotisivu.ninja/gunbug2 test to explain how to bug is now happening.

So some of the bugs are gone but there is still problem changing property between types.

**Comment by amark** [comment]
@Napzu did 3f70fa9 finally fix this!??? 

For my 4th (?) time, I bet it does!!!

**Comment by Napzu** [comment]
Nope, still not working :(

Also in my earlier testpage, ACK's are basicly not firing at all.
https://www.kotisivu.ninja/gunbug

**Comment by amark** [comment]
@Napzu said this from the gitter regarding the latest fixes/release:

"""
@amark I tested in my https://www.kotisivu.ninja/gunbug2 and https://www.kotisivu.ninja/gunbug page here is quick list of findings
now prop type change works!!! But ACK are not firing
actually when you put(ANY OBJECT) the ACK is not firing
after put(OBJ) the last OBJ ACK triggers if you change that back to some simple value put('something')
with put(OBJ) no peers get synced with new data, this propably relates to fact that the props and data I am changing back and forth has already exists?
also changing between values put( null, 'str', int, boolean ) no ACK if firing
"""

**Comment by finwo** [comment]
@amark This looks like the same behavior as #687 

I'm still working on it, but it looks like fetching the soul of where to write the data never returns (the `get` within `chain.put`), therefor the data is never marked as ready-to-write (`as.stun`), which is why the server never receives the event.

**Comment by amark** [comment]
Yeah, if `as.stun` doesn't resolve :( the write won't happen. I'm guessing logic in `.once` or in `.get(cb, true)` (which tell it to get only the soul) is not resolving properly due to debounce logic and/or checking that multiple peers have replied.

so lines like https://github.com/amark/gun/blob/master/gun.js#L1292 or https://github.com/amark/gun/blob/master/gun.js#L1666 could be giving it trouble.

 > NOTE! These are ugly lines and cannot just be deleted. Deleting them will cause the OPPOSITE problem, false-positive (or early triggering) of data not being "found" simply because a local peer or other doesn't have the data (when another slower peer might have data).

Knowing whether data exists or not, in a P2P network is.... tricky, and GUN tries to implement sensible defaults of what counts as authoritative or not - but this should be configurable for devs depending upon their needs/use cases.

Thanks @finwo for looking into these!
